src/interpreter/compiler.ts

```diff
@@ -0,0 +1,290 @@
+    jumpOnFalseInstr.instrAddress = state.instructions.length;
+    compile(node.alternate, state);
+    gotoInstr.instrAddress = state.instructions.length;
+  },
+  Constant: (node: Constant, state: CompilerState) => {
+    const loadConstantInstr = constructLoadConstantInstr(node.value);
+    state.instructions.push(loadConstantInstr);
+  },
+  ContinueStatement: (node: ContinueStatement, state: CompilerState) => {},
+  DefaultStatement: (node: DefaultStatement, state: CompilerState) => {},
+  DoWhileStatement: (node: DoWhileStatement, state: CompilerState) => {},
+  EmptyStatement: (node: EmptyStatement, state: CompilerState) => {},
+  ExpressionStatement: (node: ExpressionStatement, state: CompilerState) => {},
+  ForStatement: (node: ForStatement, state: CompilerState) => {},
+  FunctionDeclaration: (node: FunctionDeclaration, state: CompilerState) => {
+    const loadFunctionInstr = constructLoadFunctionInstr(PLACEHOLDER_ADDRESS);
+    state.instructions.push(loadFunctionInstr);
+    const gotoInstr = constructGotoInstr(PLACEHOLDER_ADDRESS);
+    state.instructions.push(gotoInstr);
+
+    loadFunctionInstr.functionInstrAddress = state.instructions.length;
+
+    // TODO: Replace param declarations when parameter list is supported
+    const functionSymbolTable = addFunctionSymbolTableEntries(
+      [],
+      node,
+      state.symbolTable
+    );
+    if (!isEmptyStatement(node.body)) {
+      node.body.items.forEach((item) => {
+        state.symbolTable = functionSymbolTable;
+        compile(item, state);
+      });
+    }
+    const teardownInstr = constructTeardownInstr();
+    state.instructions.push(teardownInstr);
+
+    gotoInstr.instrAddress = state.instructions.length;
+
+    const assignInstr = constructAssignInstr(
+      getSymbolTableEntry(node.id.name, state.symbolTable)
+    );
+    state.instructions.push(assignInstr);
+  },
+  GotoStatement: (node: GotoStatement, state: CompilerState) => {},
+  Identifier: (node: Identifier, state: CompilerState) => {
+    const loadSymbolInstr = constructLoadSymbolInstr(
+      getSymbolTableEntry(node.name, state.symbolTable)
+    );
+    state.instructions.push(loadSymbolInstr);
+  },
+  IdentifierStatement: (node: IdentifierStatement, state: CompilerState) => {},
+  IfStatement: (node: IfStatement, state: CompilerState) => {},
+  LogicalExpression: (node: LogicalExpression, state: CompilerState) => {
+    let conditionalExpression;
+    if (node.operator === '&&') {
+      conditionalExpression = constructConditionalExpression(
+        node.left,
+        node.right,
+        constructFalseConstant()
+      );
+    }
+    if (node.operator === '||') {
+      conditionalExpression = constructConditionalExpression(
+        node.left,
+        constructTrueConstant(),
+        node.right
+      );
+    }
+    if (conditionalExpression === undefined) {
+      throw new UnsupportedOperatorError(
+        node.operator,
+        UnsupportedOperatorErrorType.LOGICAL
+      );
+    }
```

Comment:
What if we convert this into a switch statement and handle the unexpected situation as the default case? Not sure if it better captures the semantics here.

Reply:
In switch, if all the cases are handled then node.operator is automatically typed to `never` so I cannot make the error message :(